docs/pip.md

```diff
@@ -0,0 +1,195 @@
+```
+
+### Specifying dependencies
+
+In addition to specifying direct dependencies, recursive dependencies also need to be explicitly
+defined for two reasons:
+
+1. Further enable reproducibility by explicitly specifying _every_ needed package
+2. Prevent the need for remote execution of `setup.py`
+
+#### pip-compile
+
+While this might be onerous to manually maintain, `pip-compile` from
+[pip-tools](https://github.com/jazzband/pip-tools) can be used to automate this process for you
+using the following procedure.
+
+1. rename `requirements.txt` to `requirements.in` (by convention)
+2. run `pip-compile requirements.in -o requirements.txt`
+
+This is the output of the above command:
+
+```ini
+#
+# This file is autogenerated by pip-compile
+# To update, run:
+#
+#    pip-compile --output-file=requirements.txt requirements.in
+#
+certifi==2020.6.20        # via requests
+chardet==3.0.4            # via requests
+git+https://github.com/containerbuildsystem/dockerfile-parse  # via -r requirements.in
+idna==2.10                # via requests
+https://github.com/containerbuildsystem/operator-manifest/archive/v0.0.3.zip  # via -r requirements.in
+requests==2.24.0          # via -r requirements.in
+ruamel.yaml.clib==0.2.2   # via ruamel.yaml
+ruamel.yaml==0.16.12      # via operator-manifest
+six==1.15.0               # via dockerfile-parse
+urllib3==1.25.10          # via requests
+```
+
+As you can see, `pip-compile` gathered all the recursive dependencies and pinned all PyPI packages.
+It did not pin external dependencies, the mechanism for doing so is specific to Cachito. You can pin
+these beforehand in the `requirements.in` file, but if any of the recursive dependencies are
+external, you may need to edit the generated file anyway.
+
+### Explicit package names
+
+Cachito needs to know the package name for all of your dependencies. For PyPI dependencies, this is
+trivial, as the name is already present in the requirements file. For external dependencies,
+resolving the name may require executing the `setup.py` file. Cachito does have a mechanism for
+extracting package metadata from `setup.py`, but it is very limited. That is why, for external
+dependencies, you will need to explicitly specify package names using one of the mechanisms that
+pip supports.
+
+a) use `@`:
+
+```
+<package-name> @ git+https://github.com/namespace/repo
+```
+
+b) use `#egg`:
+
+```
+git+https://github.com/namespace/repo#egg=<package-name>
+```
+
+Similarly to the procedure used for pinning external dependency versions, you can specify explicit
+package names in `requirements.in` to avoid having to edit the file generated by `pip-compile`.
+However, `pip-compile` seems to ignore the `@` mechanism, so using `#egg` may be preferrable.
+
+After pinning versions and specifying package names for external dependencies, the `requirements.in`
+file at the [top of this section](#requirementstxt) would look like this (commit ref and hash value
+truncated for readability):
```

Comment:
Can the hashes be truncated in Cachito? If not, I would argue that you should do something like one of these

* Put the full digest here and take the hit to readability
* Replace it with a placeholder like `<insert-full-sha-digest>`
* Make it _explicit_ i the paragraph that the full sha is needed and that the sample is not considered valid because the full sha is not used.

Reply:
and I guess something might also need to be clarified when you use the same standard above...

Reply:
Fair point, I replaced the digest and commit ref with placeholders (they must both be full length)